lib/puppet/state_machine/state.rb

```diff
@@ -0,0 +1,109 @@
+    # outcome of some query operation. Each event is associated with a transition naming the next state
+    # to enter.
+    #
+    # Transitions define a directed edge between states in a state machine. A transition is defined as the
+    # source state, the target state, and an event that causes that transition to be taken.
+    #
+    # ## Namespacing
+    #
+    # When state machines are composed individual state names need to remain unique. States can be namespaced
+    # which converts state names and transition names from a single symbol to a list of symbols.
+    #
+    class State
+
+      # @!attribute [r] name
+      #   @return [Symbol, Array<Symbol>] The name of this state. This name is used when determining and validating state machine transitions.
+      attr_reader :name
+
+      # @!attribute [r] transitions
+      #   @return [Hash<Symbol, Symbol>] A map of events and their corresponding transitions.
+      attr_reader :transitions
+      attr_reader :options
+
+      # @param name [Symbol, Array<Symbol>] A symbol or list of symbols identifying this state.
+      # @param action_cb [#call()] The action callback to invoke when entering this state. The return value of this state is passed to the event callback.
+      # @param event_cb [#call(Object)] The event callback to determine the next transition to take.
+      # @param options [Hash] A map of options indicating what kind of state this is.
+      # @param transitions [Hash<Symbol, <Symbol>>] A map of events and their corresponding transitions.
+      def initialize(name, action_cb, event_cb, options, transitions = {})
+        @name = name
+        @action_cb = action_cb
+        @event_cb = event_cb
+        @options = options
+
+        @transitions = transitions
+      end
+
+      def call
+        event(action)
+      end
+
+      def action
+        @action_cb.call
+      end
+
+      def event(result)
+        @event_cb.call(result)
+      end
+
+      def transition_for(event)
+        @transitions[event]
+      end
+
+      def error?
+        @options[:type] == :error
+      end
+
+      def final?
+        @options[:type] == :final
+      end
+
+      # Generate a new state with namespaced transitions.
+      #
+      # In order to compose multiple state machines into a single machine, the states of an
+      # individual machine need to be namespaced to prevent name collisions across different
+      # machines and identify which states originated from which of the individual machines.
+      #
+      # @param ns [Symbol]
+      # @return [Puppet::StateMachine::State]
+      def namespace(ns)
+        new_name = Array(@name).dup.unshift(ns)
+        new_transitions = Hash[@transitions.map do |(event, target)|
+          [event, Array(target).dup.unshift(ns)]
+        end]
+
+        new_options = @options.dup
+        case new_options[:type]
+        when :final
+          new_event_cb = ->(_) { :'*final*' }
```

Comment:
When we namespace a state with the intent to compose multiple state machines, we remove the `final` state type from final states and add a default event, `:'*final*'`, so that we can transition from a final state of one machine to another machine. Is this confusing? Is the addition/use of `:'*final*' too magical/confusing and if so what would be a better option?